internal/cli/cmd/cluster/grpcproxy.go

```diff
@@ -0,0 +1,208 @@
+	if !ok {
+		err := status.Errorf(codes.Internal, "reading method failed")
+		fmt.Fprintf(console.Debug(ctx), "reading method failed: %v\n", err)
+		return err
+	}
+
+	fmt.Fprintf(console.Debug(ctx), "handler %s\n", fullMethodName)
+	md, _ := metadata.FromIncomingContext(serverStream.Context())
+	outgoingCtx := metadata.NewOutgoingContext(serverStream.Context(), md.Copy())
+	backendConn, err := g.newBackendClient(outgoingCtx)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "creating backend connection failed: %v\n", err)
+		return status.Errorf(codes.Internal, "failed to connect to backend: %v", err)
+	}
+
+	clientCtx, clientCancel := context.WithCancel(outgoingCtx)
+	defer clientCancel()
+
+	clientStream, err := grpc.NewClientStream(clientCtx, clientStreamDescForProxying, backendConn, fullMethodName)
+	if err != nil {
+		fmt.Fprintf(console.Debug(ctx), "failed to create client stream: %v\n", err)
+		return status.Errorf(codes.Internal, "failed create client: %v", err)
+	}
+
+	s2cErrChan := g.proxyServerToClient(serverStream, clientStream)
+	c2sErrChan := g.proxyClientToServer(clientStream, serverStream)
+	// Make sure to close both client and server connections
+	for i := 0; i < 2; i++ {
+		select {
+		case s2cErr := <-s2cErrChan:
+			if s2cErr == io.EOF {
+				clientStream.CloseSend()
+			} else {
+				clientCancel()
+				fmt.Fprintf(console.Debug(ctx), "failed proxying s2c: %v\n", s2cErr)
+				return status.Errorf(codes.Internal, "failed proxying s2c: %v", s2cErr)
+			}
+
+		case c2sErr := <-c2sErrChan:
+			serverStream.SetTrailer(clientStream.Trailer())
+			// c2sErr will contain RPC error from client code. If not io.EOF return the RPC error as server stream error.
+			if c2sErr != io.EOF {
+				fmt.Fprintf(console.Debug(ctx), "failed proxying c2s: %v\n", c2sErr)
+				return c2sErr
+			}
+
+			return nil
+
+		case <-ctx.Done():
+			err := ctx.Err()
+			fmt.Fprintf(console.Debug(ctx), "server stream context is done: %v\n", err)
+			return err
+		}
+	}
+
+	return status.Errorf(codes.Internal, "gRPC proxy should never reach this stage.")
+}
+
+func (g *grpcProxy) proxyClientToServer(src grpc.ClientStream, dst grpc.ServerStream) chan error {
+	ret := make(chan error, 1)
+	go func() {
+		// Not really empty message. All fields are unmarshalled in Empty.unknownFields.
+		f := &emptypb.Empty{}
+		first := true
+		for {
+			if err := src.RecvMsg(f); err != nil {
+				ret <- err // this can be io.EOF which is happy case
+				break
+			}
+
+			if first {
+				first = false
+				// Server headers are only readable after first client msg is
+				// received but must be written to server stream before the first msg is sent
+				md, err := src.Header()
+				if err != nil {
+					ret <- err
+					break
+				}
+				if err := dst.SendHeader(md); err != nil {
+					ret <- err
+					break
+				}
+			}
+
+			if err := dst.SendMsg(f); err != nil {
+				ret <- err
+				break
+			}
+		}
+	}()
+	return ret
```

Comment:
There's a chance there's a path added above where we'll forget to write `err`; my advice would be to decouple the writing to the channel, from the error handling. Something like:

```
ret := make(chan error, 1)
go func() {
  if err := doProxy(src, dst); err != nil {
    ret <- err
  }
}()
return ret
```

And then have `doProxy` be a func that returns errors.